engine/src/main/scala/cromwell/engine/workflow/JobInputEvaluator.scala

```diff
@@ -0,0 +1,176 @@
+   *
+   * Where
+   *   BFN is the WDL function implementation from the backend t will run on
+   *   EFN is the WDL function implementation from the engine
+   *
+   * q, r and s will be evaluated sequentially, such that when evaluating r,
+   * the value of q has already been resolved in the appropriate scope and evaluated by the appropriate wdl functions.
+   */
+class JobInputEvaluator(workflowDescriptor: EngineWorkflowDescriptor) {
+
+  private val calls = workflowDescriptor.backendDescriptor.workflowNamespace.workflow.calls
+
+  /*
+   *  /|\ this could be very time consuming, as it potentially involves IO when using engine functions
+   * /_._\
+   */
+  def resolveAndEvaluate(jobKey: BackendJobDescriptorKey,
+                         wdlFunctions: WdlStandardLibraryFunctions,
+                         outputStore: OutputStore): Try[Map[LocallyQualifiedName, WdlValue]] = {
+    val call = jobKey.call
+    lazy val callInputsFromFile = unqualifiedInputsFromInputFile(call)
+    lazy val workflowScopedLookup = buildWorkflowScopedLookup(jobKey, outputStore) _
+
+    // Try to resolve, evaluate and coerce declarations in order
+    val inputEvaluationAttempt = call.task.declarations.foldLeft(Map.empty[LocallyQualifiedName, Try[WdlValue]])((inputs, declaration) => {
+      val name = declaration.name
+
+      // Try to resolve the declaration, and upon success evaluate the expression
+      // If the declaration is resolved but can't be evaluated this will throw an evaluation exception
+      // If it can't be resolved it's ignored and won't appear in the final input map
+      val evaluated: Option[Try[WdlValue]] = declaration.expression match {
+        // Static expression in the declaration
+        case Some(expr) => Option(expr.evaluate(buildMapBasedLookup(inputs), wdlFunctions))
+        // Expression found in the input mappings
+        case None if call.inputMappings.contains(name) => Option(call.inputMappings(name).evaluate(workflowScopedLookup, wdlFunctions))
+        // Expression found in the input file
+        case None if callInputsFromFile.contains(name) => Option(Success(callInputsFromFile(name)))
+        // Expression can't be found
+        case _ => None
+      }
+
+      // Leave out unresolved declarations
+      evaluated match {
+        case Some(value) =>
+          val coercedValue = value flatMap declaration.wdlType.coerceRawValue
+          inputs + ((name, coercedValue))
+        case None => inputs
+      }
+    })
+
+    TryUtil.sequenceMap(inputEvaluationAttempt, s"Input evaluation for Call ${call.fullyQualifiedName} failed")
+  }
+
+  private def splitFqn(fullyQualifiedName: FullyQualifiedName): (String, String) = {
+    val lastIndex = fullyQualifiedName.lastIndexOf(".")
+    (fullyQualifiedName.substring(0, lastIndex), fullyQualifiedName.substring(lastIndex + 1))
+  }
+
+  // Split inputs map (= evaluated workflow declarations + coerced json inputs) into [init\.*].last
```

Comment:
You're right about the regexp, and it is confusing yes I should put an example instead